docs/metrics-data-model.md

```diff
@@ -0,0 +1,108 @@
+# k6 Time Series
+
+We want to introduce the time series concept to k6 for getting the benefit of efficiently identifying the combination between a metric and the relative tags.
+
+A summary of the benefits it brings across the platform:
+
+* It allows executing faster lookups and comparisons by a single value, the ID (hash) of the series.
+* It reduces the storage by allocating just one time a possible combination (the time series) and referencing it in all other places. 
+* It allows aggregating data by time series.
+* It enables easier detection of high cardinality issues just by counting the entire unique data set of generated time series.
+* It enables time series based outputs to simplify the integration's logic (e.g. Prometheus).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* Time Series (metric, tags)
+* Sample (timestamp, value, time series ref)
+* Tag - (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+```go
+// This is the current k6 Metric polished from the Thresholds dependencies.
+type Metric struct {
+	name     string
+	type     MetricType
+	contains ValueType
+}
+
+type Tag struct {
+    Key, Value string
+}
+
+type TimeSeries struct {
+	id uint64 // hash(metric+tags)
+	metricName string
+	tags   []Tag
+	keys map[string]struct{}
```

Comment:
```suggestion
	keys map[string]int
```

or something like that, to allow for direct lookup? :thinking:

Reply:
mmm...I thinking if we should do one step more at this point. My idea of tags map was something `map[keySepValue]struct{}` for having the maximum efficiency on `Contains` but it wouldn't support the fetch by `Tag.Key`.

So we could consider this data structure for supporting the single `GetTag(key)`:

```go
tags map[key]stuct{
    value string
    valhash uint64
}
```
In this way, we could support most of the operations with one data structure. The unique missing operation is returning the sorted []Tag for example for `GetTags()`, I see two potential solution:
* store just the keys in a sorted slice
* add another field like `order/index` in the `struct` so we know in which specific place set the `Tag` item.  

```go
func (ts TimeSeries) GetTags() []Tag {
    s := make([]Tag, len(ts.tags))
    for k, data := range {
       s[data.index] = Tag{Key: k, Value: data.Value}
    }
    return s
}
```

The list of operations we have potentially consider in comparisons:

```go
func (TimeSeries) Contains(subset []Tag) bool
func (TimeSeries) HasKey(string) bool 
func (TimeSeries) GetTag(string) Tag
func (TimeSeries) GetTags() []Tag
```

@na-- Thoughts?

Reply:
Thinking better on this, it doesn't really work. Because on `Contains` we don't have the relative `tag.valhash` to compare. So I think, we can just support the `GetTag` by binary search :thinking:

Reply:
I am not sure I understand all of your concerns, or maybe I'm missing something :confused: Please correct me if I'm wrong, but we basically need to be able to efficiently access the tags both in a sorted sequential order and in a random access order (i.e. lookup by their key). So, the basic implementation choices are these:
1. Store the tag keys and values in a sorted list (`[]Tag`) for efficient sequential access and have an index (`map[string]int`) by the key for efficient random access 
2. Store the tag keys and values in a map (`map[string]string`) for the efficient random access and have a sorted list of keys as a `[]string` slice for the sequential access.
3. Something weirder

The first approach seems preferable to me, since the second one involves an extra map per tag access when iterating across all tags sequentially, and anything else seems unnecessary over-engineering at this early point. :man_shrugging: 

Given that these time series objects should be measured in thousands at most, and not millions, I don't see why we'd want anything more complicated than this? You probably won't save all that much memory usage if you don't have the index and use binary search, for example, since I think the strings for the keys will be reused between the slice and the map :man_shrugging: 

Besides, we might be verging in bike-shedding territory here :slightly_smiling_face: . The actual implementation should be internal to the `TimeSeries` object, right? So I don't think it matters very much implementation you pick, we should be able to change it at any time in the future without anything breaking, right?